src/storage/cache/s3fifo.rs

```diff
@@ -0,0 +1,213 @@
+use hashbrown::HashMap;
+use std::cmp::min;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::sync::atomic::AtomicU8;
+use std::sync::atomic::Ordering::Relaxed;
+
+use crate::storage::cache::queue::Queue;
+
+const MAX_FREQUENCY_LIMIT: u8 = 3;
+
+struct Entry<K, V> {
+    key: K,
+    value: V,
+    freq: AtomicU8,
+}
+
+impl<K, V> Entry<K, V> {
+    pub fn new(key: K, value: V) -> Self {
+        Self {
+            key,
+            value,
+            freq: AtomicU8::new(0),
+        }
+    }
+}
+
+/// Cache is an implementation of "S3-FIFO" from "FIFO Queues are ALL You Need for Cache Eviction" by
+/// Juncheng Yang, et al: https://jasony.me/publication/sosp23-s3fifo.pdf
+pub struct Cache<K, V>
+where
+    K: PartialEq + Eq + Hash + Clone + Debug,
+    V: Clone,
+{
+    min_eviction_size: usize,
+    max_ghost_size: usize,
+    max_cache_size: usize,
+    small: Queue<Entry<K, V>>,
+    main: Queue<Entry<K, V>>,
+    ghost: Queue<K>,
+    table: HashMap<K, Entry<K, V>>,
+}
+
+impl<K, V> Cache<K, V>
+where
+    K: PartialEq + Eq + Hash + Clone + Debug,
+    V: Clone,
+{
+    pub fn new(max_cache_size: usize) -> Self {
```

Comment:
Have you considered taking a `NonZeroUsize` to avoid having to assert below?

Reply:
Thanks for the suggestion; I have not. I just have one problem with the NonZeroUsize, and that is having to import it from a separate std crate. Doesn't seem friendly as an API to import a different type just to delegate the assert. But if you say this is a very common pattern in Rust and most people use it, I'd be happy to use that here.

Reply:
Using `NonZeroUsize` improves the type safety and it clearly communicates to the caller that this function takes only non-zero arguments without even needing to read the documentation or code. I love self documenting APIs like that. I know a `usize` looks better aesthetically but I will pick the extra steps of constructing a `NonZeroUsize` over a potential runtime panic anytime.

A `usize` would be fine if the function gave zero a special meaning, not when it panics. We could also document the panic but I don't think that's the best user experience. Panics should be avoided whenever possible. In this case, it's easy to avoid it.

Reply:
I agree